### ISSUE TITLE
Lambda type cannot be inferred when type context includes a callable with a callable parameter

### ISSUE THREAD

**Issue by erictraut** [python/mypy#16523]
This was filed as a [pyright bug](https://github.com/microsoft/pyright/issues/6326). I noticed that mypy likewise doesn't handle this case, and I couldn't find any open issues that cover this.

Mypy playground [link](https://mypy-play.net/?mypy=latest&python=3.11&gist=b6663db6173e8d353874fa3a2eedb51c)

```python
from typing import Callable, Generic, TypeVar

T = TypeVar("T")

def func1(callable: Callable[[Callable[[T], T]], T]) -> T:
    return callable(lambda x: x)

v1 = func1(lambda a: a(0))
reveal_type(v1)  # Should be int

v2 = func1(lambda a: a(""))
reveal_type(v2)  # should be str

class A(Generic[T]):
    def __init__(self, value: T) -> None:
        self.value = value

def func2(callable: Callable[[type[A[T]]], A[T]]) -> T:
    return callable(A).value

v3 = func2(lambda A: A(0))
reveal_type(v3)  # Should be int

v4 = func2(lambda A: A(""))
reveal_type(v4)  # Should be str
```

Expected Results: No Errors and revealed types as indicated in the comments

Actual Results: Multiple errors ("Need type annotation for ...").
